presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBuffer.java

```diff
@@ -110,6 +112,26 @@
      */
     void fail();
 
+    /**
+     * Notify buffer that no more pages will be added for the given lifespan.
+     * Any future calls to enqueue a page of that lifespan are ignored.
+     * @see OutputBuffer#setNoMorePages()
+     */
+    void setNoMorePagesForLifespan(Lifespan lifespan);
+
+    /**
+     * Register a callback which get called once a buffer is finished for a lifespan.
+     * This method should be called exactly once before any data is produced to the output buffer.
+     */
+    void registerLifespanCompletionCallback(Consumer<Lifespan> callback);
+
+    /**
+     * A buffer is finished for the given lifespan once no-more-pages has been set for that lifespan
+     * and all pages has been consumed.
+     * @see OutputBuffer#isFinished()
+     */
+    boolean isFinishedForLifespan(Lifespan lifespan);
```

Comment:
Having two `isFinished` method variants is confusing/complex for the users of this API. Do you think it makes sense to have a single `isFinished(Lifespan)` method and implement it in a way that for `Lifespan.TASK_WIDE` input it behaves the same as the current `isFinished` method? I think this comment also applies to the `setNoMorePages` method. In general it's good to keep the API surface area small and keep it simple.

Reply:
I was also thinking about this. But it seems that it would get messier if we had only a single `isFinish(Lifespan)`.

Reply:
It's a trade-off between the simplicity of the API vs. the complexity of the internal implementation. We can still have these two `isFinished` variants in the internal implementation, but have a simpler public API.

Reply:
Yeah, I also thought about this, but I found it's a really tough call. The question is really about whether `TASK_WIDE` can be used to represent overall execution. 

Unfortunately, all `lifespan` related APIs and usages in Presto is not that well polished and considered. 

My personal understanding is, `TASK_WIDE` is a lifespan only makes sense for ungrouped execution. So, under the context of grouped execution, `TASK_WIDE` is an invalid lifespan and shouldn't be used to indicate overall execution. And we need dedicated API (`isFinished` vs. `isFinishedForLifespan` to distinguish them). 

This aligns with how `TaskSource` handles `noMorePages` signal, for lifespan and for overal execution. 

(think about `TaskSource` update as a RPC call for `setNoMoreSplitsForLifespan` and `setNoMoreSplits`:
https://github.com/prestodb/presto/blob/bfeb574b995c812801ec9d67ef34ce6cc43b6bb4/presto-main/src/main/java/com/facebook/presto/execution/TaskSource.java#L31-L32

and 

https://github.com/prestodb/presto/blob/bfeb574b995c812801ec9d67ef34ce6cc43b6bb4/presto-main/src/main/java/com/facebook/presto/execution/TaskSource.java#L64-L74


Does this sounds reasonable?